### PR TITLE
add new plugin: emulators/qemu-guest-agent

### LINKED ISSUE #1586: Add QEMU Guest Agent package
As already discussed in this thread https://forum.opnsense.org/index.php?topic=5860.15 It would be  nice to add this packto OPNSense ports.

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=232949

### PR DESCRIPTION
**Pull request by fraenki**
closes #1586

### CODE REVIEW

**Comment by fraenki** [comment]
@fichtner Feel free to review if you like, but it's OK if you don't have time for this :)

File: emulators/qemu-guest-agent/Makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+PLUGIN_NAME=		qemu-guest-agent
+PLUGIN_VERSION=		1.0
+PLUGIN_COMMENT=		QEMU Guest Agent for OPNsense
+PLUGIN_PLUGIN_DEPENDS=		qemu-guest-agent
```
Comment (fichtner) [inline emulators/qemu-guest-agent/Makefile:4]:
```suggestion
PLUGIN_DEPENDS=		qemu-guest-agent
```

File: emulators/qemu-guest-agent/Makefile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+PLUGIN_NAME=		qemu-guest-agent
+PLUGIN_VERSION=		1.0
```
Comment (fichtner) [inline emulators/qemu-guest-agent/Makefile:2]:
```suggestion
PLUGIN_VERSION=		0.1
PLUGIN_DEVEL=		yes
```

File: emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+# Check if the kernel module exists
+KERNMOD='virtio_console'
+if [ -e /boot/kernel/${KERNMOD}.ko ]; then
+    # Load module
+    kldload $KERNMOD 2>&1
```
Comment (fichtner) [inline emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh:7]: Might already be loaded. I think kldstat -m can test for this?

File: emulators/qemu-guest-agent/src/opnsense/service/templates/OPNsense/QemuGuestAgent/rc.conf.d [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{# Default setting is enabled, so that no GUI interaction is required. #}
+{% if not (helpers.exists('OPNsense.QemuGuestAgent.general.LogDebug')) or OPNsense.QemuGuestAgent.general.Enabled|default("0") != "0" %}
+{%   set optional_flags = [] %}
+{%   do optional_flags.append('-d -l /var/log/qemu-ga.log') %}
+{%   if helpers.exists('OPNsense.QemuGuestAgent.general.LogDebug') and OPNsense.QemuGuestAgent.general.LogDebug|default("0") == "1" %}
```
Comment (AdSchellevis) [inline emulators/qemu-guest-agent/src/opnsense/service/templates/OPNsense/QemuGuestAgent/rc.conf.d:5]: we do have a shorthand for this nowadays, see the example below:

```
{%    if not helpers.empty('OPNsense.QemuGuestAgent.general.LogDebug') %}
```

Comment (fraenki) [reply emulators/qemu-guest-agent/src/opnsense/service/templates/OPNsense/QemuGuestAgent/rc.conf.d:5]: Good to know, thanks!

File: emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh (hunk 3 above)
Comment (fraenki) [reply emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh:7]: Yes, in theory, but it does not work for whatever reason...

```
root@opnsense: # kldstat 
Id Refs Address                Size Name
 1   35 0xffffffff80200000  26904e8 kernel
 2    1 0xffffffff82891000    10250 carp.ko
 3    1 0xffffffff828a2000     f998 if_bridge.ko
 4    2 0xffffffff828b2000     72a8 bridgestp.ko
 5    1 0xffffffff828ba000     3e78 if_enc.ko
 6    1 0xffffffff828be000     b1c0 if_gre.ko
 7    1 0xffffffff828ca000    16008 if_lagg.ko
 8    1 0xffffffff828e1000     8b60 if_tap.ko
 9    3 0xffffffff828ea000    583b8 pf.ko
10    1 0xffffffff82943000     2af8 pflog.ko
11    1 0xffffffff82946000     ebd0 pfsync.ko
12    1 0xffffffff82a12000     2668 intpm.ko
13    1 0xffffffff82a15000      b50 smbus.ko
14    1 0xffffffff82a16000     18a0 uhid.ko
15    1 0xffffffff82a18000     1aa0 wmt.ko
16    1 0xffffffff82a1a000     2aac virtio_console.ko

root@opnsense: # kldstat -m virtio_console
kldstat: can't find module virtio_console: No such file or directory

root@opnsense: # kldload virtio_console.ko
kldload: can't load virtio_console.ko: module already loaded or in kernel
```

...but it works for other modules:

```
root@opnsense: # kldstat -m carp
Id  Refs Name
  1    1 carp
```

Comment (fraenki) [reply emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh:7]: I could still do something like `if ! kldstat | grep virtio_console; then  kldload virtio_console; fi`, but that's not more efficient, so nothing to gain here I guess.

Comment (fichtner) [reply emulators/qemu-guest-agent/src/opnsense/scripts/OPNsense/QemuGuestAgent/setup.sh:7]: Weird, but alright then 😊

File: emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    $services[] = array(
+        'description' => gettext('QEMU Guest Agent'),
+        'pidfile' => '/var/run/qemu-ga.pid',
+        'configd' => array(
+            'restart' => array('qemuguestagent restart'),
+            'start' => array('qemuguestagent start'),
+            'stop' => array('qemuguestagent stop'),
+        ),
+        'name' => 'qemuguestagent',
```
Comment (fichtner) [inline emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc:57]: Normally this is the binary‘s name

Comment (fraenki) [reply emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc:57]: Ah, I didn't know that. Will rename it to `qemu-ga`then.

Comment (fraenki) [reply emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc:57]: hm, if I rename it to `qemu-ga`, then it does not show up in `System: Diagnostics: Services` anymore.

Comment (fichtner) [reply emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc:57]: I‘m not aware of such limitations. Syntax error maybe? It doesn’t read the file on syntax error.

Comment (fraenki) [reply emulators/qemu-guest-agent/src/etc/inc/plugins.inc.d/qemuguestagent.inc:57]: Silly me. I've tested with a brand new config, so `qemuguestagent_enabled()` returned `false` and as a result the service did not show up in the GUI :)

File: emulators/qemu-guest-agent/src/opnsense/mvc/app/models/OPNsense/QemuGuestAgent/QemuGuestAgent.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+/**
+ *    Copyright (C) 2021 Deciso B.V.
+ *    Copyright (C) 2015 Deciso B.V.
```
Comment (fichtner) [inline emulators/qemu-guest-agent/src/opnsense/mvc/app/models/OPNsense/QemuGuestAgent/QemuGuestAgent.php:5]: Should this have been your copyright line?

Comment (fichtner) [reply emulators/qemu-guest-agent/src/opnsense/mvc/app/models/OPNsense/QemuGuestAgent/QemuGuestAgent.php:5]: Not this one, but the one above.

File: emulators/qemu-guest-agent/src/opnsense/mvc/app/controllers/OPNsense/QemuGuestAgent/IndexController.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+/**
+ *    Copyright (C) 2021 Deciso B.V.
```
[the remark this answers is not in the document]
Comment (fraenki) [reply emulators/qemu-guest-agent/src/opnsense/mvc/app/controllers/OPNsense/QemuGuestAgent/IndexController.php]: Fixed, thanks.

File: emulators/qemu-guest-agent/src/opnsense/mvc/app/models/OPNsense/QemuGuestAgent/QemuGuestAgent.php (hunk 6 above)
Comment (fraenki) [reply emulators/qemu-guest-agent/src/opnsense/mvc/app/models/OPNsense/QemuGuestAgent/QemuGuestAgent.php:5]: Fixed, thanks.

**Review by fichtner — APPROVED** [review]
Ok to merge. This goes in first, then tools after a short ports tree test. Thanks a lot!

**Comment by drewhemm** [comment]
Hi, any way to test a preview release of this?

**Comment by fraenki** [comment]
@drewhemm Install os-qemu-guest-agent-devel from the command line:

```
pkg install os-qemu-guest-agent-devel
```

(Be sure to use the latest version of OPNsense before trying to install it.)

**Comment by fraenki** [comment]
The next release of OPNsense will include the final version of the qemu-guest-agent plugin (#2357).

**Pull request merged**